pkg/scheduler/ugm/queue_tracker.go

```diff
@@ -316,104 +312,132 @@ func (qt *QueueTracker) IsQueuePathTrackedCompletely(queuePath string) bool {
-	if immediateChildQueueName != common.Empty {
-		if childUt, ok := qt.childQueueTrackers[immediateChildQueueName]; ok {
-			return childUt.IsUnlinkRequired(childQueuePath)
+func (qt *QueueTracker) IsUnlinkRequired(hierarchy []string) bool {
+	// depth first: all the way to the leaf, ignore if not exists
+	// more than 1 in the slice means we need to recurse down
+	if len(hierarchy) > 1 {
+		childName := hierarchy[1]
+		if qt.childQueueTrackers[childName] != nil {
+			return qt.childQueueTrackers[childName].IsUnlinkRequired(hierarchy[1:])
 		}
-	}
-	if queuePath == configs.RootQueue || queuePath == qt.queueName {
-		if len(qt.runningApplications) == 0 {
-			log.Log(log.SchedUGM).Debug("Is Unlink Required?",
-				zap.String("queue path", queuePath),
-				zap.Int("no. of applications", len(qt.runningApplications)))
-			return true
+	} else if len(hierarchy) == 1 {
+		// reach end of hierarchy
+		if hierarchy[0] == configs.RootQueue || hierarchy[0] == qt.queueName {
+			if len(qt.runningApplications) == 0 {
+				log.Log(log.SchedUGM).Debug("Is Unlink Required?",
+					zap.String("queue path", qt.queuePath),
+					zap.Int("no. of applications", len(qt.runningApplications)))
+				return true
+			}
 		}
 	}
 	return false
 }
 
 // UnlinkQT Traverse queue path upto the end queue. If end queue has any more child queue trackers,
 // then goes upto each child queue and removes the linkage with its immediate parent
-func (qt *QueueTracker) UnlinkQT(queuePath string) bool {
+func (qt *QueueTracker) UnlinkQT(hierarchy []string) bool {
 	log.Log(log.SchedUGM).Debug("Unlinking current queue tracker from its parent",
 		zap.String("current queue ", qt.queueName),
-		zap.String("queue path", queuePath),
+		zap.String("queue path", qt.queuePath),
+		zap.Strings("hierarchy", hierarchy),
 		zap.Int("no. of child queue trackers", len(qt.childQueueTrackers)))
-	childQueuePath, immediateChildQueueName := getChildQueuePath(queuePath)
-
-	if childQueuePath == common.Empty && len(qt.childQueueTrackers) > 0 {
-		for qName := range qt.childQueueTrackers {
-			qt.UnlinkQT(qt.queueName + configs.DOT + qName)
-		}
-	}
-
-	if childQueuePath != common.Empty {
-		if qt.childQueueTrackers[immediateChildQueueName] != nil {
-			unlink := qt.childQueueTrackers[immediateChildQueueName].UnlinkQT(childQueuePath)
-			if unlink {
-				delete(qt.childQueueTrackers, immediateChildQueueName)
+	// depth first: all the way to the leaf, ignore if not exists
+	// more than 1 in the slice means we need to recurse down
+	if len(hierarchy) > 1 {
+		childName := hierarchy[1]
+		if qt.childQueueTrackers[childName] != nil {
+			if qt.childQueueTrackers[childName].UnlinkQT(hierarchy[1:]) {
+				delete(qt.childQueueTrackers, childName)
 			}
 		}
+	} else if len(hierarchy) <= 1 {
+		// reach end of hierarchy, unlink all queues under this queue
+		qt.unlinkAllQT()
 	}
+
 	if len(qt.runningApplications) == 0 && len(qt.childQueueTrackers) == 0 {
 		return true
 	}
 	return false
 }
```

Comment:
Can we rewrite this method to avoid newly introduced method `unlinkAllQT()` ?

```
func (qt *QueueTracker) UnlinkQT(hierarchy []string) bool {
	log.Log(log.SchedUGM).Debug("Unlinking current queue tracker from its parent",
		zap.String("current queue ", qt.queueName),
		zap.String("queue path", qt.queuePath),
		zap.Strings("hierarchy", hierarchy),
		zap.Int("no. of child queue trackers", len(qt.childQueueTrackers)))
	// depth first: all the way to the leaf, ignore if not exists
	// more than 1 in the slice means we need to recurse down
	if len(hierarchy) > 1 {
		childName := hierarchy[1]
		if qt.childQueueTrackers[childName] != nil {
			if qt.childQueueTrackers[childName].UnlinkQT(hierarchy[1:]) {
				delete(qt.childQueueTrackers, childName)
			}
		}
	} else if len(hierarchy) <= 1 {
		// reach end of hierarchy, unlink all queues under this queue
		for childName, childQT := range qt.childQueueTrackers {
			if childQT.UnlinkQT(strings.Split(childName, configs.DOT)) {
				delete(qt.childQueueTrackers, childName)
			}
		}
	}

	if len(qt.runningApplications) == 0 && len(qt.childQueueTrackers) == 0 {
		return true
	}
	return false
}

```

Can we try this out?

Reply:
Agree. I think this can work. The only different part is that we get child queue **name** from `childQueueTrackers`, not child queue **path**, so we don't need to do `strings.Split` on it.